File: classified/packages/plugin-plugin-manager/src/services/pluginRegistryService.ts
import { elizaLogger } from '@elizaos/core';
import type { ExecException } from 'node:child_process';
import { PluginMetadata } from '../types';

const API_SERVICE_URL = process.env.ELIZAOS_API_URL || 'http://localhost:3000/api';
const API_KEY = process.env.ELIZAOS_API_KEY || '';

export interface PluginSearchResult {
  id?: string;
  name: string;
  description: string;
  score?: number;
  tags?: string[];
  features?: string[];
  requiredConfig?: string[];
  version?: string;
  npmPackage?: string;
  repository?: string;
  relevantSection?: string;
}

export interface PluginDetails {
  id?: string;
  name: string;
  description: string;
  readme?: string;
  capabilitiesSummary?: string;
  repository?: string;
  version?: string;
  tags?: string[];
  features?: string[];
  requiredConfig?: string[];
  dependencies?: Record<string, string>;
  dependedBy?: string[];
  npmPackage?: string;
  installCount?: number;
  searchCount?: number;
  categories?: string[];
  lastIndexed?: string;
  createdAt?: string;
  updatedAt?: string;
}

export interface CloneResult {
  success: boolean;
  error?: string;
  pluginName?: string;
  localPath?: string;
  hasTests?: boolean;
  dependencies?: Record<string, string>;
}

interface ApiResponse<T> {
  success: boolean;
  data?: T;
  error?: string;
}

async function apiFetch<T>(endpoint: string, options: RequestInit = {}): Promise<ApiResponse<T>> {
  const url = `${API_SERVICE_URL}${endpoint}`;
  const headers = {
    'Content-Type': 'application/json',
    Authorization: `Bearer ${API_KEY}`,
    ...options.headers,
  };

  try {
    const response = await fetch(url, { ...options, headers });
    if (!response.ok) {
      const errorBody = await response.text();
      elizaLogger.error(
        `[pluginRegistryService] API request to ${url} failed with status ${response.status}: ${errorBody}`
      );
      throw new Error(`API request failed: ${response.statusText}`);
    }
    return response.json() as Promise<ApiResponse<T>>;
  } catch (error) {
    elizaLogger.error(`[pluginRegistryService] Network error calling ${url}:`, error);
    throw error;
  }
}

export async function searchPluginsByContent(query: string): Promise<PluginSearchResult[]> {
  elizaLogger.info(`[pluginRegistryService] Searching for plugins matching: ${query}`);
  try {
    const data = await apiFetch<PluginSearchResult[]>('/plugins/search', {
      method: 'POST',
      body: JSON.stringify({ query, limit: 10 }),
    });
    return data.data || [];
  } catch (error) {
    return [];
  }
}

export async function getPluginDetails(name: string): Promise<PluginMetadata | null> {
  elizaLogger.info(`[pluginRegistryService] Getting details for plugin: ${name}`);
  try {
    const data = await apiFetch<PluginMetadata>(`/plugins/${encodeURIComponent(name)}`, {
      method: 'GET',
    });
    return data.data || null;
  } catch (error) {
    return null;
  }
}

export async function getAllPlugins(): Promise<PluginMetadata[]> {
  elizaLogger.info('[pluginRegistryService] Getting all plugins from registry');
  try {
    const data = await apiFetch<PluginMetadata[]>('/plugins', {
      method: 'GET',
    });
    return data.data || [];
  } catch (error) {
    return [];
  }
}

export async function getPluginsByCategory(category: string): Promise<PluginMetadata[]> {
  elizaLogger.info(`[pluginRegistryService] Getting plugins for category: ${category}`);
  try {
    const data = await apiFetch<PluginMetadata[]>(
      `/plugins?category=${encodeURIComponent(category)}`,
      {
        method: 'GET',
      }
    );
    return data.data || [];
  } catch (error) {
    return [];
  }
}

export async function clonePlugin(pluginName: string): Promise<CloneResult> {
  elizaLogger.info(`[pluginRegistryService] Cloning plugin: ${pluginName}`);

  const plugin = await getPluginDetails(pluginName);
  if (!plugin || !plugin.repository) {
    return {
      success: false,
      error: `Plugin "${pluginName}" not found in registry or has no repository`,
    };
  }

  try {
    const path = await import('node:path');
    const fs = await import('node:fs/promises');
    const { exec } = await import('node:child_process');
    const { promisify } = await import('node:util');
    const execAsync = promisify(exec);

    const cloneDir = path.join(
      process.cwd(),
      'cloned-plugins',
      plugin.name.replace('@elizaos/', '')
    );

    await fs.mkdir(cloneDir, { recursive: true });
    await execAsync(`git clone ${plugin.repository} ${cloneDir}`);

    const packageJsonPath = path.join(cloneDir, 'package.json');
    const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf-8'));
    const hasTests = !!(packageJson.scripts?.test || packageJson.devDependencies?.vitest);

    return {
      success: true,
      pluginName: plugin.name,
      localPath: cloneDir,
      hasTests,
      dependencies: packageJson.dependencies || {},
    };
  } catch (error) {
    elizaLogger.error('[pluginRegistryService] Clone failed:', error);
    const execError = error as ExecException;
    return {
      success: false,
      error: execError.stderr || execError.message || 'Unknown error during clone',
    };
  }
}

File: excpp/classified/packages/plugin-plugin-manager/src/services/pluginRegistryService.cpp
#include "pluginRegistryService.hpp"
#include <iostream>
#include <stdexcept>

namespace elizaos {

std::future<std::vector<PluginSearchResult>> searchPluginsByContent(const std::string& query) {
    // NOTE: Auto-converted from TypeScript - may need refinement

    "[pluginRegistryService] Searching for plugins matching: " + std::to_string(query)
    try {
        const auto data = apiFetch<PluginSearchResult[]>('/plugins/search', {;
            method: 'POST',
            body: JSON.stringify({ query, limit: 10 }),
            });
            return data.data || [];
            } catch (error) {
                return [];
            }

}

std::future<std::optional<PluginMetadata>> getPluginDetails(const std::string& name) {
    // NOTE: Auto-converted from TypeScript - may need refinement

    "[pluginRegistryService] Getting details for plugin: " + std::to_string(name)
    try {
        const auto data = "/plugins/" + std::to_string(encodeURIComponent(name));
            method: 'GET',
            });
            return data.data || nullptr;
            } catch (error) {
                return nullptr;
            }

}

std::future<std::vector<PluginMetadata>> getAllPlugins() {
    // NOTE: Auto-converted from TypeScript - may need refinement

    elizaLogger.info('[pluginRegistryService] Getting all plugins from registry');
    try {
        const auto data = apiFetch<PluginMetadata[]>('/plugins', {;
            method: 'GET',
            });
            return data.data || [];
            } catch (error) {
                return [];
            }

}

std::future<std::vector<PluginMetadata>> getPluginsByCategory(const std::string& category) {
    // NOTE: Auto-converted from TypeScript - may need refinement

    "[pluginRegistryService] Getting plugins for category: " + std::to_string(category)
    try {
        const auto data = apiFetch<PluginMetadata[]>(;
        "/plugins?category=" + std::to_string(encodeURIComponent(category))
        {
            method: 'GET',
        }
        );
        return data.data || [];
        } catch (error) {
            return [];
        }

}

std::future<CloneResult> clonePlugin(const std::string& pluginName) {
    // NOTE: Auto-converted from TypeScript - may need refinement

    "[pluginRegistryService] Cloning plugin: " + std::to_string(pluginName)

    const auto plugin = getPluginDetails(pluginName);
    if (!plugin || !plugin.repository) {
        return {
            success: false,
            "Plugin "" + std::to_string(pluginName) + "" not found in registry or has no repository"
            };
        }

        try {
            const auto path = import('node:path');
            const auto fs = import('node:fs/promises');
            const auto { exec } = import('node:child_process');
            const auto { promisify } = import('node:util');
            const auto execAsync = promisify(exec);

            const auto cloneDir = path.join(;
            process.cwd(),
            'cloned-plugins',
            plugin.name.replace('@elizaos/', '');
            );

            fs.mkdir(cloneDir, { recursive: true });
            "git clone " + std::to_string(plugin.repository) + " " + std::to_string(cloneDir);

            const auto packageJsonPath = path.join(cloneDir, 'package.json');
            const auto packageJson = JSON.parse(fs.readFile(packageJsonPath, 'utf-8'));
            const auto hasTests = !!(packageJson.scripts.test || packageJson.devDependencies.vitest);

            return {
                success: true,
                pluginName: plugin.name,
                localPath: cloneDir,
                hasTests,
                dependencies: packageJson.dependencies || {},
                };
                } catch (error) {
                    elizaLogger.error('[pluginRegistryService] Clone failed:', error);
                    const auto execError = error;
                    return {
                        success: false,
                        error: execError.stderr || execError.message || 'Unknown error during clone',
                        };
                    }

}

} // namespace elizaos
